### PR TITLE
Examples update

### PR DESCRIPTION
**Pull request by eddiebergman**
# Features
* Makes the quickstart guide runnable by [sphinx-gallry](https://sphinx-gallery.github.io/stable/index.html).
* Makes examples also runnable by sphinx-gallery.
* Adds make command `make doc-with-examples` which will generate examples outputs with doc, `make doc` will just generate text. Also include the API building portion and the linkchecking portion of doc building.

# Important Note
* Examples are only runnable in `sphinx-gallery` with python `> 3.9`, see [Version Issues]().

# Major changes
* `quickstart.rst` was converted to a `quickstart_example.py` to enable sphinx-gallery.
* Examples are seperated into categories.
* Examples renamed from `*.py` to `*_example.py`, unless file name was `*_func.py`, in which case it was unchanged.

# Minor Changes
* Flake 8 fixes for spear_qcp as there were a lot of complaints running `pre-commit`.

# To build and view
```bash
$ cd SMAC3
$ make clean
$ make doc-with-examples
$ xdg-open ./doc/build/html/index.html  # Or firefox, google-chrome, etc.. 
```

# Does not
* Does not improve the code or text of the examples except for minor changes in `quickstart_example.py`

# Version issues
The examples are successfully by `sphinx-gallery` with `python 3.9.5`, but this is not the case with `python 3.8.6`. Nothing in the [python version notes](https://docs.python.org/release/3.9.6/whatsnew/3.9.html) indicates why this might be the case.

As this is only likely to be run by us for generating the doc and publishing it, I think this will have to do for now.
I imagine this might be revisited in the future, or else live with this issue until support for `python < 3.9` is dropped.

# Issues
* The linkchecking portion of `make doc` complains about many links yet manual checking seems to show all of them working correctly.

* 6 examples fail to run using `python < 3.9.0` for unknown import errors, the log of which is given below 
<details>
<summary>`make doc`</summary>

```
make -C doc html
make[1]: Entering directory '/home/skantify/code/SMAC3/doc'
sphinx-build -b html -d build/doctrees   . build/html
Running Sphinx v4.1.1
loading pickled environment... done
[autosummary] generating autosummary for: api.rst, basic_usage.rst, contact.rst, examples/SMAC4BO/SMAC4BO_mlp_example.rst, examples/SMAC4BO/SMAC4BO_rosenbrock_example.rst, examples/SMAC4BO/SMAC4BO_sgd_instances_example.rst, examples/SMAC4BO/sg_execution_times.rst, examples/SMAC4HPO/SMAC4HPO_acq_rosenbrock_example.rst, examples/SMAC4HPO/SMAC4HPO_gradientboosting_example.rst, examples/SMAC4HPO/SMAC4HPO_rosenbrock_example.rst, ..., main_options.rst, manual.rst, options.rst, psmac.rst, scenario_options.rst, smac_examples.rst, smac_options.rst, tae.rst, usage_recomendation.rst, validation.rst
generating gallery...
WARNING: /home/skantify/code/SMAC3/examples/quickstart/quickstart_example.py failed to execute correctly: Traceback (most recent call last):                                                    
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/quickstart/quickstart_example.py", line 56, in <module>
    from smac.facade.func_facade import fmin_smac
ModuleNotFoundError: No module named 'smac.facade'


generating gallery for examples/function_minimization... [100%] fmin_rosenbrock_example.py                                                                                                      
WARNING: /home/skantify/code/SMAC3/examples/parallel/fmin_rosenbrock_parallel_example.py failed to execute correctly: Traceback (most recent call last):                                        
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/parallel/fmin_rosenbrock_parallel_example.py", line 12, in <module>
    from smac.intensification.simple_intensifier import SimpleIntensifier
ModuleNotFoundError: No module named 'smac.intensification.simple_intensifier'

WARNING: /home/skantify/code/SMAC3/examples/parallel/parallel_sh_mlp_example.py failed to execute correctly: Traceback (most recent call last):                                                 
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/parallel/parallel_sh_mlp_example.py", line 22, in <module>
    from smac.facade.roar_facade import ROAR
ModuleNotFoundError: No module named 'smac.facade.roar_facade'


generating gallery for examples/SMAC4HPO... [100%] SMAC4HPO_svm_example.py                                                                                                                      
WARNING: /home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_rosenbrock_example.py failed to execute correctly: Traceback (most recent call last):                                               
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_rosenbrock_example.py", line 14, in <module>
    from smac.facade.smac_bo_facade import SMAC4BO
ModuleNotFoundError: No module named 'smac.facade.smac_bo_facade'

WARNING: /home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_sgd_instances_example.py failed to execute correctly: Traceback (most recent call last):                                            
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_sgd_instances_example.py", line 30, in <module>
    from smac.facade.smac_mf_facade import SMAC4MF
ModuleNotFoundError: No module named 'smac.facade.smac_mf_facade'

generating gallery for examples/SMAC4BO... [100%] SMAC4BO_mlp_example.py                                                                                                                        
generating gallery for examples/SMAC4MF... [100%] SMAC4MF_mlp.py                                                                                                                                
WARNING: /home/skantify/code/SMAC3/examples/hyperband/hyperband_mlp_example.py failed to execute correctly: Traceback (most recent call last):                                                  
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/hyperband/hyperband_mlp_example.py", line 25, in <module>
    from smac.facade.hyperband_facade import HB4AC
ModuleNotFoundError: No module named 'smac.facade.hyperband_facade'


computation time summary:
    - ../examples/hyperband/hyperband_mlp_example.py:                 0.01 sec   0.0 MB
    - ../examples/SMAC4BO/SMAC4BO_sgd_instances_example.py:           0.01 sec   0.0 MB
    - ../examples/parallel/parallel_sh_mlp_example.py:                0.00 sec   0.0 MB
    - ../examples/quickstart/quickstart_example.py:                   0.00 sec   0.0 MB
    - ../examples/SMAC4BO/SMAC4BO_rosenbrock_example.py:              0.00 sec   0.0 MB
    - ../examples/parallel/fmin_rosenbrock_parallel_example.py:       0.00 sec   0.0 MB
    - ../examples/SMAC4BO/SMAC4BO_mlp_example.py:                     0.00 sec   0.0 MB
    - ../examples/SMAC4HPO/SMAC4HPO_acq_rosenbrock_example.py:        0.00 sec   0.0 MB
    - ../examples/SMAC4HPO/SMAC4HPO_gradientboosting_example.py:      0.00 sec   0.0 MB
    - ../examples/SMAC4HPO/SMAC4HPO_rosenbrock_example.py:            0.00 sec   0.0 MB
    - ../examples/SMAC4HPO/SMAC4HPO_svm_example.py:                   0.00 sec   0.0 MB
    - ../examples/SMAC4MF/SMAC4MF_mlp.py:                             0.00 sec   0.0 MB
    - ../examples/SMAC4MF/SMAC4MF_sgd_instances.py:                   0.00 sec   0.0 MB
    - ../examples/function_minimization/fmin_rosenbrock_example.py:   0.00 sec   0.0 MB
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 11 source files that are out of date
updating environment: 0 added, 15 changed, 0 removed
reading sources... [100%] smac_options                                                                                                                                                          
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.callbacks'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.configspace'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.epm'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.facade'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.initial_design'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.intensification'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.optimizer'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.runhistory'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.scenario'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.stats'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.tae'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.utils'
/home/skantify/code/SMAC3/doc/api.rst:9: WARNING: toctree contains reference to nonexisting document 'apidoc/smac.smac_cli'
/home/skantify/code/SMAC3/doc/examples/index.rst:22: WARNING: duplicate label sphx_glr_examples_quickstart, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:22: WARNING: duplicate label quickstart, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:59: WARNING: duplicate label sphx_glr_examples_function_minimization, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:59: WARNING: duplicate label fmin_examples, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:94: WARNING: duplicate label sphx_glr_examples_parallel, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:94: WARNING: duplicate label parallel_examples, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:149: WARNING: duplicate label sphx_glr_examples_smac4hpo, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:149: WARNING: duplicate label smac4hpo_examples, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:242: WARNING: duplicate label sphx_glr_examples_smac4bo, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:317: WARNING: duplicate label sphx_glr_examples_smac4mf, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:317: WARNING: duplicate label smac4mf_examples, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:372: WARNING: duplicate label sphx_glr_examples_hyperband, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
/home/skantify/code/SMAC3/doc/examples/index.rst:372: WARNING: duplicate label hyperband_examples, other instance in /home/skantify/code/SMAC3/doc/smac_examples.rst
examples/index.rst:22: WARNING: duplicate label sphx_glr_examples_quickstart, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:22: WARNING: duplicate label quickstart, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:59: WARNING: duplicate label sphx_glr_examples_function_minimization, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:59: WARNING: duplicate label fmin_examples, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:94: WARNING: duplicate label sphx_glr_examples_parallel, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:94: WARNING: duplicate label parallel_examples, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:149: WARNING: duplicate label sphx_glr_examples_smac4hpo, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:149: WARNING: duplicate label smac4hpo_examples, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:242: WARNING: duplicate label sphx_glr_examples_smac4bo, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:317: WARNING: duplicate label sphx_glr_examples_smac4mf, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:317: WARNING: duplicate label smac4mf_examples, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:372: WARNING: duplicate label sphx_glr_examples_hyperband, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
examples/index.rst:372: WARNING: duplicate label hyperband_examples, other instance in /home/skantify/code/SMAC3/doc/examples/index.rst
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] smac_options                                                                                                                                                           
generating indices... genindex py-modindex done
writing additional pages... search done
copying images... [100%] examples/hyperband/images/thumb/sphx_glr_hyperband_mlp_example_thumb.png                                                                                               
copying downloadable files... [100%] examples/quickstart/quickstart_example.ipynb                                                                                                               
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 45 warnings.

The HTML pages are in build/html.

Sphinx-gallery successfully executed 0 out of 6 files subselected by:

    gallery_conf["filename_pattern"] = '.*example.py$|.*tutorial.py$'
    gallery_conf["ignore_pattern"]   = '.*_func.py'

after excluding 6 files that had previously been run (based on MD5).


Extension error (sphinx_gallery.gen_gallery):
Handler <function summarize_failing_examples at 0x7f8bea4158b0> for event 'build-finished' threw an exception (exception: Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/home/skantify/code/SMAC3/examples/quickstart/quickstart_example.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/quickstart/quickstart_example.py", line 56, in <module>
    from smac.facade.func_facade import fmin_smac
ModuleNotFoundError: No module named 'smac.facade'


/home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_sgd_instances_example.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_sgd_instances_example.py", line 30, in <module>
    from smac.facade.smac_mf_facade import SMAC4MF
ModuleNotFoundError: No module named 'smac.facade.smac_mf_facade'


/home/skantify/code/SMAC3/examples/parallel/parallel_sh_mlp_example.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/parallel/parallel_sh_mlp_example.py", line 22, in <module>
    from smac.facade.roar_facade import ROAR
ModuleNotFoundError: No module named 'smac.facade.roar_facade'


/home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_rosenbrock_example.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/SMAC4BO/SMAC4BO_rosenbrock_example.py", line 14, in <module>
    from smac.facade.smac_bo_facade import SMAC4BO
ModuleNotFoundError: No module named 'smac.facade.smac_bo_facade'


/home/skantify/code/SMAC3/examples/parallel/fmin_rosenbrock_parallel_example.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/parallel/fmin_rosenbrock_parallel_example.py", line 12, in <module>
    from smac.intensification.simple_intensifier import SimpleIntensifier
ModuleNotFoundError: No module named 'smac.intensification.simple_intensifier'


/home/skantify/code/SMAC3/examples/hyperband/hyperband_mlp_example.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/skantify/code/SMAC3/.venv/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/skantify/code/SMAC3/examples/hyperband/hyperband_mlp_example.py", line 25, in <module>
    from smac.facade.hyperband_facade import HB4AC
ModuleNotFoundError: No module named 'smac.facade.hyperband_facade'


-------------------------------------------------------------------------------)
make[1]: *** [Makefile:59: html] Error 2
make[1]: Leaving directory '/home/skantify/code/SMAC3/doc'
make: *** [Makefile:12: doc-with-examples] Error 2
```

</details>

### CODE REVIEW

**Comment by renesass** [comment]
Gonna have a look into that later

**Comment by dengdifan** [comment]
1. We renamed BOHB4HPO to SMAC4MF in a previous PR #738, could you please do the same to all  BOHB related files?
2. We deleted the example for RF in #722, maybe you could do the same?

**Comment by eddiebergman** [comment]
I will get around to it in the second half of the day hopefully, do you guys have a major release for the paper planned soon?

**Comment by renesass** [comment]
actually, I don't like how the examples are sorted. Can we maybe create subfolders? E.g. 
SMAC4HPO_rf_example.py -> SMAC4HPO/rf.py
hyperband_mlp_example.py -> HB/mlp.py

**Comment by renesass** [comment]
the whole spear_qcp seems way to complex to me. We should consider moving it to the general examples.

**Comment by benjamc** [comment]
> 
> 
> actually, I don't like how the examples are sorted. Can we maybe create subfolders? E.g.
> SMAC4HPO_rf_example.py -> SMAC4HPO/rf.py
> hyperband_mlp_example.py -> HB/mlp.py

Good idea. Would look cleaner.

**Comment by benjamc** [comment]
> 
> 
> the whole spear_qcp seems way to complex to me. Way too much files, too. Are they even needed and called? We should consider moving it to the general examples.

Moving it to general sounds good, and it is good to have this more complex example. Maybe we could add a really small readme saying that we want to optimize spear qcp and what it is (I didn't know:D)  and with which script/file to start.

**Review by renesass — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 1]
```diff
@@ -32,6 +32,12 @@ doc/main_options.rst
 doc/scenario_options.rst
 doc/smac_options.rst
 doc/apidoc
+doc/examples
+doc/quickstart
+
+# Generated from running Sphinx-Gallery
+examples/general/smac3-output_*
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply .gitignore:39]: Good idea

File: examples/general/parallel_sh_mlp_example.py [hunk 2]
```diff
@@ -32,7 +32,9 @@
 # --------------------------------------------------------------
 import os
 import sys
-sys.path.append(os.path.join(os.path.dirname(__file__)))
+# https://sphinx-gallery.github.io/stable/faq.html#why-is-file-not-defined-what-can-i-use
+# sys.path.append(os.path.join(os.path.dirname(__file__)))
+sys.path.append(os.getcwd())
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply examples/general/parallel_sh_mlp_example.py:37]: I left the original as a comment along with the link to indicate why. Anyone viewing the example might consider `os.cwd()` a bit odd to see. 

When Sphinx-Gallery runs an example, the variable `__file__` is undefined. It runs the file from the file's source directory so `os.cwd()` accomplishes the same task and is the recommended solution by them. 

https://sphinx-gallery.github.io/stable/faq.html#why-is-file-not-defined-what-can-i-use

**Comment by eddiebergman** [comment]
> actually, I don't like how the examples are sorted. Can we maybe create subfolders? E.g.
> SMAC4HPO_rf_example.py -> SMAC4HPO/rf.py
> hyperband_mlp_example.py -> HB/mlp.py

I began moving them to their own subfolders as I like the idea too but the issue is that a lot of examples rely on `mlp_from_cfg_func.py`.

The line `from mlp_from_cfg_func import mlp_from_cfg` that exists in most of them only work if the importing file and `mlp_from_cfg_func.py` exist in the same directory. It is possible to just add more to the path but it obscures the examples a bit in my opinion. In general, I'm not really a fan of multi-file examples as it introduces more areas to go wrong.

For now, I will try to give the examples a module structure, introducing `__init__.py` and see if I can get relative imports working.
i.e.
```
/quickstart/
/general/
---- /subfolder1/
---- ---- __init__.py
---- ---- example1.py
---- /subfolder2/
---- ---- __init__.py
---- ---- example2.py
---- __init__.py
---- mlp_from_cfg_func.py
```
And then in the examples
```python
# example1.py
from ..mlp_from_cfg_func import mlp_from_cfg
```

The alternative as suggested is to add to the parent path in each example
```python
sys.path.append(os.getcwd()) # Add directory the example is in
sys.path.append(os.path.join('..', os.getcwd()) # Add parent directory
```
This will however break if any folder restructuring is done and won't work if a user downloads the examples and runs the example themselves outside the directory.
```sh
$ cwd
> ..../general
$ python subfolder1/example1.py
> Something along the lines of `mlp_from_cfg_func` can't be imported
```

**Comment by eddiebergman** [comment]
Scratch that, the module method didn't work. For now a simple solution is just add the parent directory and account for whether it is run by a user or by sphinx gallery (in which case `__file__` does not exist.

```python
# https://sphinx-gallery.github.io/stable/faq.html#why-is-file-not-defined-what-can-i-use
cwd = os.path.dirname(__file__) if __file__ else os.getcwd()
sys.path.append(os.path.join(cwd, os.pardir))  # mlp_from_cfg_func lives one dir up
from mlp_from_cfg_func import mlp_from_cfg  # noqa: E402
```

**Comment by eddiebergman** [comment]
Actually I noticed [SMAC4MF_mlp.py](https://github.com/automl/SMAC3/blob/development/examples/SMAC4MF_mlp.py) has the `mlp_from_cfg` function directly in the example. This removes the external file dependancy and I think this is likely the better option at the cost of making the example longer.

**Comment by eddiebergman** [comment]
There are still 6 failing examples but I can not solve them, they are all imports in only 'some' examples.

**Comment by eddiebergman** [comment]
There should also probably be tests for checking if all examples execute correctly for the future, I'm busy until next wednesday and I'm not sure when this PR is required for the papers. I can raise an issue for it if you like?

**Comment by eddiebergman** [comment]
Examples with `sphinx-gallery` are only run successfully with `python 3.9`, can't identify why it would matter

**Comment by dengdifan** [comment]
Under Manual -> Basic Usage -> Usage in python, the link to SVM-example, ExecuteTAFuncDict and ExexuteTAFuncArray are broken.
Under Manual -> Options and file formats, link to SVM-example and SPEAR example also broken 
Under Manual ->Target Algorithm Evaluator, almost all link break.
Same for Manual -> Validation

**Comment by eddiebergman** [comment]
As updated in the PR main comment, the linkcheck functionality gives a lot of false negatives. Through manual checking they all seem to work as intended. I would suggest in the future, once this release if published and over, switching how referencing is done to instead use the internal 'anchors' in each `rst` file rather than reference `html` files that will be generated.

I wrote a small example here for any future work on fixing up documentation or the false negative errors generated by `linkcheck` with sphinx
```bash
$ cd doc; make linkcheck
```

### Examples for link changing
```rst
# file_one.rst

.. _my_anchor:

.. _explicit_title_anchor:
Title
=====
```

```rst
# file_two.rst

Here I want to :ref:`another anchor in another file <my_anchor>` which
will work as long as `my_anchor` is present in any file during the build.

... as opposed to how it's `currently done <file_two.html#my_anchor>`_
which will fail if any reorganization of the docs are done such as moving `file_one` to `../file_one`.

Also using the title such as `file_one.html#Title` will fail if someone want's to change the Title displayed yet
this :ref:`one <explicit_title_anchor>` will not
```

**Review by renesass — APPROVED** [review]

**Pull request merged**